> FYI, on my side I get these failing tests, I believe there might be a small difference between our envs. We can always update the expected sentence later in case they fail on the daily CI (which probably will be the case). Happy also to add the missing test in a follow up PR.
> <img alt="Screenshot 2023-06-27 at 16 08 55" width="1314" src="https://user-images.githubusercontent.com/49240599/249179643-418abf8c-3408-49b6-8212-5e4e75e5f284.png">

Aha, it's been stable for me so far, but I can see that happening. If it's any help I'm running this on an RTX 4090 and `torch==2.1.0.dev20230603+cu121`.

> Also one test is failing for 4bit:
> 
> ```shell
> FAILED tests/bnb/test_4bit.py::Bnb4BitGPT2Test::test_memory_footprint - AttributeError: 'GPT2MLP' object has no attribute 'dense_4h_to_h'
> ```
> 
> Could quickly address a fix? 🙏 After that we should be ready to merge

Nice catch! I have a fix in mind that should also remove most of the int8 test code I added, so I'll get that in asap.

--- a/tests/bnb/test_4bit.py
+++ b/tests/bnb/test_4bit.py
@@ -39,6 +39,12 @@ from transformers.testing_utils import (
 from transformers.utils.versions import importlib_metadata
 
 
+def get_some_linear_layer(model):
+    if model.config.model_type == "gpt2":
+        return model.transformer.h[0].mlp.c_fc
+    return model.transformer.h[0].mlp.dense_4h_to_h
+
+
 if is_torch_available():
     import torch
     import torch.nn as nn
@@ -136,7 +142,8 @@ class Bnb4BitTest(Base4bitTest):
         mem_4bit = self.model_4bit.get_memory_footprint()
 
         self.assertAlmostEqual(mem_fp16 / mem_4bit, self.EXPECTED_RELATIVE_DIFFERENCE)
-        self.assertTrue(self.model_4bit.transformer.h[0].mlp.dense_4h_to_h.weight.__class__ == Params4bit)
+        linear = get_some_linear_layer(self.model_4bit)
+        self.assertTrue(linear.weight.__class__ == Params4bit)
 
     def test_linear_are_4bit(self):
         r"""

--- a/tests/bnb/test_mixed_int8.py
+++ b/tests/bnb/test_mixed_int8.py
@@ -41,6 +41,12 @@ from transformers.testing_utils import (
 from transformers.utils.versions import importlib_metadata
 
 
+def get_some_linear_layer(model):
+    if model.config.model_type == "gpt2":
+        return model.transformer.h[0].mlp.c_fc
+    return model.transformer.h[0].mlp.dense_4h_to_h
+
+
 if is_accelerate_available():
     from accelerate import PartialState
     from accelerate.logging import get_logger
@@ -142,7 +148,7 @@ class MixedInt8Test(BaseMixedInt8Test):
         mem_8bit = self.model_8bit.get_memory_footprint()
 
         self.assertAlmostEqual(mem_fp16 / mem_8bit, self.EXPECTED_RELATIVE_DIFFERENCE)
-        self.assertTrue(self.model_8bit.transformer.h[0].mlp.dense_4h_to_h.weight.__class__ == Int8Params)
+        self.assertTrue(get_some_linear_layer(self.model_8bit).weight.__class__ == Int8Params)
 
     def test_linear_are_8bit(self):
         r"""
@@ -292,8 +298,9 @@ class MixedInt8Test(BaseMixedInt8Test):
 
             model_from_saved = AutoModelForCausalLM.from_pretrained(tmpdirname, load_in_8bit=True, device_map="auto")
 
-            self.assertTrue(model_from_saved.transformer.h[0].mlp.dense_4h_to_h.weight.__class__ == Int8Params)
-            self.assertTrue(hasattr(model_from_saved.transformer.h[0].mlp.dense_4h_to_h.weight, "SCB"))
+            linear = get_some_linear_layer(model_from_saved)
+            self.assertTrue(linear.weight.__class__ == Int8Params)
+            self.assertTrue(hasattr(linear.weight, "SCB"))
 
             # generate
             encoded_input = self.tokenizer(self.input_text, return_tensors="pt")
@@ -318,8 +325,9 @@ class MixedInt8Test(BaseMixedInt8Test):
 
             model_from_saved = AutoModelForCausalLM.from_pretrained(tmpdirname)
 
-            self.assertTrue(model_from_saved.transformer.h[0].mlp.dense_4h_to_h.weight.__class__ == Int8Params)
-            self.assertTrue(hasattr(model_from_saved.transformer.h[0].mlp.dense_4h_to_h.weight, "SCB"))
+            linear = get_some_linear_layer(model_from_saved)
+            self.assertTrue(linear.weight.__class__ == Int8Params)
+            self.assertTrue(hasattr(linear.weight, "SCB"))
 
             # generate
             encoded_input = self.tokenizer(self.input_text, return_tensors="pt")
@@ -339,8 +347,9 @@ class MixedInt8Test(BaseMixedInt8Test):
 
         model = AutoModelForCausalLM.from_pretrained(model_id)
 
-        self.assertTrue(model.transformer.h[0].mlp.dense_4h_to_h.weight.__class__ == Int8Params)
-        self.assertTrue(hasattr(model.transformer.h[0].mlp.dense_4h_to_h.weight, "SCB"))
+        linear = get_some_linear_layer(model)
+        self.assertTrue(linear.weight.__class__ == Int8Params)
+        self.assertTrue(hasattr(linear.weight, "SCB"))
 
         # generate
         encoded_input = self.tokenizer(self.input_text, return_tensors="pt")
@@ -752,74 +761,8 @@ class MixedInt8TestTraining(BaseMixedInt8Test):
 
 class MixedInt8GPT2Test(MixedInt8Test):
     model_name = "gpt2-xl"
-    MLP_FC_NAME = "c_fc"
     EXPECTED_RELATIVE_DIFFERENCE = 1.8720077507258357
     EXPECTED_OUTPUT = "Hello my name is John Doe, and I am a member of the"
-
-    def test_memory_footprint(self):
-        r"""
-        A simple test to check if the model conversion has been done correctly by checking on the
-        memory footprint of the converted model and the class type of the linear layers of the converted models
-        """
-        from bitsandbytes.nn import Int8Params
-
-        mem_fp16 = self.model_fp16.get_memory_footprint()
-        mem_8bit = self.model_8bit.get_memory_footprint()
-
-        self.assertAlmostEqual(mem_fp16 / mem_8bit, self.EXPECTED_RELATIVE_DIFFERENCE)
-        self.assertTrue(self.model_8bit.transformer.h[0].mlp.c_fc.weight.__class__ == Int8Params)
-
-    def test_int8_serialization(self):
-        r"""
-        Test whether it is possible to serialize a model in 8-bit.
-        """
-        from bitsandbytes.nn import Int8Params
-
-        with tempfile.TemporaryDirectory() as tmpdirname:
-            self.model_8bit.save_pretrained(tmpdirname)
-
-            # check that the file `quantization_config` is present
-            config = AutoConfig.from_pretrained(tmpdirname)
-            self.assertTrue(hasattr(config, "quantization_config"))
-
-            model_from_saved = AutoModelForCausalLM.from_pretrained(tmpdirname, load_in_8bit=True, device_map="auto")
-
-            self.assertTrue(model_from_saved.transformer.h[0].mlp.c_fc.weight.__class__ == Int8Params)
-            self.assertTrue(hasattr(model_from_saved.transformer.h[0].mlp.c_fc.weight, "SCB"))
-
-            # generate
-            encoded_input = self.tokenizer(self.input_text, return_tensors="pt")
-            output_sequences = model_from_saved.generate(input_ids=encoded_input["input_ids"].to(0), max_new_tokens=10)
-
-            self.assertEqual(
-                self.tokenizer.decode(output_sequences[0], skip_special_tokens=True), self.EXPECTED_OUTPUT
-            )
-
-    def test_int8_serialization_sharded(self):
-        r"""
-        Test whether it is possible to serialize a model in 8-bit - sharded version.
-        """
-        from bitsandbytes.nn import Int8Params
-
-        with tempfile.TemporaryDirectory() as tmpdirname:
-            self.model_8bit.save_pretrained(tmpdirname, max_shard_size="200MB")
-
-            # check that the file `quantization_config` is present
-            config = AutoConfig.from_pretrained(tmpdirname)
-            self.assertTrue(hasattr(config, "quantization_config"))
-
-            model_from_saved = AutoModelForCausalLM.from_pretrained(tmpdirname)
-
-            self.assertTrue(model_from_saved.transformer.h[0].mlp.c_fc.weight.__class__ == Int8Params)
-            self.assertTrue(hasattr(model_from_saved.transformer.h[0].mlp.c_fc.weight, "SCB"))
-
-            # generate
-            encoded_input = self.tokenizer(self.input_text, return_tensors="pt")
-            output_sequences = model_from_saved.generate(input_ids=encoded_input["input_ids"].to(0), max_new_tokens=10)
-
-            self.assertEqual(
-                self.tokenizer.decode(output_sequences[0], skip_special_tokens=True), self.EXPECTED_OUTPUT
-            )
 
     def test_int8_from_pretrained(self):
         # TODO: Test loading quantized gpt2 model from the hub.

--- a/tests/bnb/test_mixed_int8.py
+++ b/tests/bnb/test_mixed_int8.py
@@ -765,5 +765,5 @@ class MixedInt8GPT2Test(MixedInt8Test):
     EXPECTED_OUTPUT = "Hello my name is John Doe, and I am a member of the"
 
     def test_int8_from_pretrained(self):
-        # TODO: Test loading quantized gpt2 model from the hub.
+        # TODO @younesbelkada: Test loading quantized gpt2 model from the hub.
         pass

